src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java

```diff
@@ -270,68 +250,107 @@ private Argument findArgumentByName(String name, List<Argument> arguments) {
-        GraphQLType graphQLType = inlineFragment.getTypeCondition() != null
-                ? TypeFromAST.getTypeFromAST(getValidationContext().getSchema(), inlineFragment.getTypeCondition())
-                : parentType;
-        collectFields(fieldMap, inlineFragment.getSelectionSet(), graphQLType, visitedFragmentSpreads);
+
+    private boolean sameType(GraphQLType type1, GraphQLType type2) {
+        if (type1 == null || type2 == null) {
+            return true;
+        }
+        return type1.equals(type2);
     }
 
-    private void collectFieldsForField(Map<String, List<FieldAndType>> fieldMap, GraphQLType parentType, Field field) {
-        String responseName = field.getResultKey();
-        if (!fieldMap.containsKey(responseName)) {
-            fieldMap.put(responseName, new ArrayList<>());
+
+    private static class FieldAndType {
+        final Field field;
+        final GraphQLType graphQLType;
+        final GraphQLType parentType;
+
+        public FieldAndType(Field field, GraphQLType graphQLType, GraphQLType parentType) {
+            this.field = field;
+            this.graphQLType = graphQLType;
+            this.parentType = parentType;
         }
-        GraphQLOutputType fieldType = null;
-        GraphQLUnmodifiedType unwrappedParent = unwrapAll(parentType);
-        if (unwrappedParent instanceof GraphQLFieldsContainer) {
-            GraphQLFieldsContainer fieldsContainer = (GraphQLFieldsContainer) unwrappedParent;
-            GraphQLFieldDefinition fieldDefinition = getVisibleFieldDefinition(fieldsContainer, field);
-            fieldType = fieldDefinition != null ? fieldDefinition.getType() : null;
+
+        @Override
+        public String toString() {
+            return "FieldAndType{" +
+                    "field=" + field +
+                    ", graphQLType=" + graphQLType +
+                    ", parentType=" + parentType +
+                    '}';
         }
-        fieldMap.get(responseName).add(new FieldAndType(field, fieldType, parentType));
-    }
 
-    private GraphQLFieldDefinition getVisibleFieldDefinition(GraphQLFieldsContainer fieldsContainer, Field field) {
-        return getValidationContext().getSchema().getCodeRegistry().getFieldVisibility().getFieldDefinition(fieldsContainer, field.getName());
-    }
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
 
-    private static class FieldPair {
-        final Field field1;
-        final Field field2;
+            FieldAndType that = (FieldAndType) o;
 
-        public FieldPair(Field field1, Field field2) {
-            this.field1 = field1;
-            this.field2 = field2;
+            return field != null ? field.equals(that.field) : that.field == null;
```

Comment:
This .equals some suspect.

I mean you may only ever compared them into terms of AST Field equality - but the class itself has 3 fields

I just want to make sure this is intended

Reply:
Yes that is intended: the AST field is unique and actually determines the types, meaning we only need to compare the fields.